src/Lib.P2P/MultiAddressExtensions.cs

```diff
@@ -0,0 +1,147 @@
+        }
+
+        /// <summary>
+        ///   Get all the addresses for the specified <see cref="MultiAddress"/>.
+        /// </summary>
+        /// <param name="multiaddress">
+        ///   The multiaddress to resolve.
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's result
+        ///   is a sequence of possible multiaddresses.
+        /// </returns>
+        /// <exception cref="SocketException">
+        ///   The host name cannot be resolved.
+        /// </exception>
+        /// <remarks>
+        ///   When the <see cref="NetworkProtocol.Name"/> starts with "dns", then a DNS
+        ///   lookup is performed to get all the IP addresses for the host name.  "dn4" and "dns6"
+        ///   will filter the result for IPv4 and IPV6 addresses.
+        ///   <para>
+        ///   When the <see cref="NetworkProtocol.Name"/> is "http" or "https", then
+        ///   a "tcp/80" or "tcp/443" is respectively added.
+        ///   </para>
+        /// </remarks>
+        public static async Task<List<MultiAddress>> ResolveAsync(this MultiAddress multiaddress,
+            CancellationToken cancel = default)
+        {
+            var list = new List<MultiAddress>();
+
+            // HTTP
+            var i = multiaddress.Protocols.FindIndex(ma => ma.Name == "http");
+            if (i >= 0 && !multiaddress.Protocols.Any(p => p.Name == "tcp"))
+            {
+                multiaddress = multiaddress.Clone();
+                multiaddress.Protocols.InsertRange(i + 1, _http.Protocols);
+            }
+
+            // HTTPS
+            i = multiaddress.Protocols.FindIndex(ma => ma.Name == "https");
+            if (i >= 0 && !multiaddress.Protocols.Any(p => p.Name == "tcp"))
+            {
+                multiaddress = multiaddress.Clone();
+                multiaddress.Protocols.InsertRange(i + 1, _https.Protocols);
+            }
+
+            // DNS*
+            i = multiaddress.Protocols.FindIndex(ma => ma.Name.StartsWith("dns"));
+            if (i < 0)
+            {
+                list.Add(multiaddress);
+                return list;
+            }
+
+            var protocolName = multiaddress.Protocols[i].Name;
+            var host = multiaddress.Protocols[i].Value;
+
+            // TODO: Don't use DNS, but use the IPFS Engine DNS resolver.
+            // This will not then expose the domain name in plain text.
+            // We also, then get to specify if A and/or AAAA records are needed.
+            var addresses = (await Dns.GetHostAddressesAsync(host).ConfigureAwait(false))
+               .Where(a => _supportedDnsAddressFamilies.ContainsKey(a.AddressFamily))
+               .Where(a =>
+                    protocolName == "dns" ||
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)